fold/fold.go

```diff
@@ -0,0 +1,879 @@
+			return deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+		}
+	}
+
+	if start > 0 && end < len(foldContext.seq)-1 {
+		// it's internal
+		foldEnergy, ok := foldContext.energies.nearestNeighbors[paired]
+		if !ok {
+			foldEnergy = foldContext.energies.internalMismatches[paired]
+		}
+		enthalpyHDifference, entropySDifference := foldEnergy.enthalpyH, foldEnergy.entropyS
+		return deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+	}
+	if start == 0 && end == len(foldContext.seq)-1 {
+		// it's terminal
+		foldEnergy, ok := foldContext.energies.nearestNeighbors[paired]
+		if !ok {
+			foldEnergy = foldContext.energies.internalMismatches[paired]
+		}
+		enthalpyHDifference, entropySDifference := foldEnergy.enthalpyH, foldEnergy.entropyS
+		return deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+	}
+
+	if start > 0 && end == len(foldContext.seq)-1 {
+		// it's dangling on left
+		foldEnergy, ok := foldContext.energies.nearestNeighbors[paired]
+		if !ok {
+			foldEnergy = foldContext.energies.internalMismatches[paired]
+		}
+		enthalpyHDifference, entropySDifference := foldEnergy.enthalpyH, foldEnergy.entropyS
+		dG := deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+
+		pairDanglingEnds := fmt.Sprintf("%c%c/.%c", foldContext.seq[start-1], foldContext.seq[start], foldContext.seq[end])
+		if foldEnergy, ok := foldContext.energies.danglingEnds[pairDanglingEnds]; ok {
+			enthalpyHDifference, entropySDifference := foldEnergy.enthalpyH, foldEnergy.entropyS
+			dG += deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+		}
+		return dG
+	}
+
+	if start == 0 && end < len(foldContext.seq)-1 {
+		// it's dangling on right
+		foldEnergy, ok := foldContext.energies.nearestNeighbors[paired]
+		if !ok {
+			foldEnergy = foldContext.energies.internalMismatches[paired]
+		}
+		enthalpyHDifference, entropySDifference := foldEnergy.enthalpyH, foldEnergy.entropyS
+		dG := deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+
+		pairDanglingEnds := fmt.Sprintf(".%c/%c%c", +foldContext.seq[start], foldContext.seq[end+1], foldContext.seq[end])
+		if foldEnergy, ok := foldContext.energies.danglingEnds[pairDanglingEnds]; ok {
+			enthalpyHDifference, entropySDifference := foldEnergy.enthalpyH, foldEnergy.entropyS
+			dG += deltaG(enthalpyHDifference, entropySDifference, foldContext.temp)
+			return dG
+		}
+	}
+	return 0
+}
+
+// hairpin calculates the free energy of a hairpin.
+// Args:
+//
+//		start:  The index of start of hairpin
+//		end:  The index of end of hairpin
+//	 foldContext: The FoldingContext for this sequence
+//
+// Returns the free energy increment from the hairpin structure
+func hairpin(start, end int, foldContext context) (float64, error) {
+	if end-start < minLenForStruct {
+		return math.Inf(1), nil
+	}
+
+	hairpinSeq := foldContext.seq[start : end+1]
+	hairpinLength := len(hairpinSeq) - 2
```

Comment:
why -2?

Reply:
The closing BP? But I din't find a good explanation.